### PR TITLE
supports openharmony for v3.17.2

### PR DESCRIPTION
**Pull request by zhaoyang57**

### CODE REVIEW

File: extensions/Particle3D/PU/CCPUMaterialManager.cpp [hunk 1]
```diff
@@ -159,7 +164,17 @@ bool PUMaterialCache::loadMaterialsFromSearchPaths( const std::string &fileFolde
...
     }
     AAssetDir_close(dir);
-
+#elif (CC_TARGET_PLATFORM == CC_PLATFORM_OHOS)
+    FileUtilsOhos *fileUtilsOhos = dynamic_cast<FileUtilsOhos*>(FileUtils::getInstance());
+    std::vector<std::string> files = fileUtilsOhos -> listFiles(fileFolder);
+    for (auto fileName : files) {
+        if (FileUtils::getInstance()->getFileExtension(fileName) == ".material")
+        {
+            std::string fullpath = std::string(fileName);
```
Comment (qiuguohua) [inline extensions/Particle3D/PU/CCPUMaterialManager.cpp]: The type of fileName is std::string?

File: cocos/audio/AudioEngine.cpp [hunk 2]
```diff
@@ -152,6 +154,9 @@ class AudioEngine::AudioEngineThreadPool
 
 void AudioEngine::end()
 {
+#if (CC_TARGET_PLATFORM == CC_PLATFORM_OHOS)
+    stopAll();
```
Comment (dumganhar) [inline cocos/audio/AudioEngine.cpp]: Why does OHOS need to invoke `stopAll` in `end` but other platforms don't?

File: cmake/Modules/CocosBuildSet.cmake [hunk 3]
```diff
@@ -24,8 +24,8 @@ message(STATUS "ENGINE_BINARY_PATH:" ${ENGINE_BINARY_PATH})
 
 # the default behavior of build module
 option(DEBUG_MODE "Debug or Release?" ON)
-option(BUILD_LUA_LIBS "Build lua libraries" OFF)
-option(BUILD_JS_LIBS "Build js libraries" OFF)
+option(BUILD_LUA_LIBS "Build lua libraries" ON)
+option(BUILD_JS_LIBS "Build js libraries" ON)
```
Comment (qiuguohua) [inline cmake/Modules/CocosBuildSet.cmake]: Other platforms turn it on by default? Why not use the configuration?

File: cocos/platform/ohos/CCApplication-ohos.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
+#if CC_TARGET_PLATFORM == CC_PLATFORM_OHOS
```
Comment (qiuguohua) [inline cocos/platform/ohos/CCApplication-ohos.cpp:1]: Why do we need this macro, this file should be specific to openharmony?

File: cocos/platform/ohos/CCApplication-ohos.h [hunk 5]
```diff
@@ -0,0 +1,89 @@
+#ifndef __CC_APPLICATION_OHOS_H__
+#define __CC_APPLICATION_OHOS_H__
+
+#include "platform/CCPlatformConfig.h"
+#if CC_TARGET_PLATFORM == CC_PLATFORM_OHOS
```
Comment (qiuguohua) [inline cocos/platform/ohos/CCApplication-ohos.h:5]: Why do we need this macro, this file should be specific to openharmony?

File: cocos/audio/AudioEngine.cpp [hunk 6]
```diff
@@ -378,6 +404,22 @@ void AudioEngine::uncache(const std::string &filePath)
     auto audioIDsIter = _audioPathIDMap.find(filePath);
     if (audioIDsIter != _audioPathIDMap.end())
     {
+	#if (CC_TARGET_PLATFORM == CC_PLATFORM_OHOS)
+        auto lst =  _audioPathIDMap[filePath];
```
Comment (minggo) [inline cocos/audio/AudioEngine.cpp]: - audioIDsIter contains the result, so can just use it
- and if still need to use it like this, i think it is better to use const reference to improve performance

File: cocos/audio/AudioEngine.cpp [hunk 7]
```diff
@@ -378,6 +404,22 @@ void AudioEngine::uncache(const std::string &filePath)
...
+        for (auto it = lst.begin() ; it != lst.end(); ++it) {
```
Comment (minggo) [inline cocos/audio/AudioEngine.cpp]: How about use `for in` loop?

File: cocos/audio/include/AudioEngine.h [hunk 8]
```diff
@@ -175,6 +175,11 @@ class EXPORT_DLL AudioEngine
     /** Pause all playing audio instances. */
     static void pauseAll();
 
+    #if CC_TARGET_PLATFORM == CC_PLATFORM_OHOS
+    static void onEnterBackground();
+    static void onEnterForeground();
+    #endif
```
Comment (minggo) [inline cocos/audio/include/AudioEngine.h]: Why need these functions?

File: cocos/base/ccConfig.h [hunk 9]
```diff
@@ -162,7 +162,11 @@ THE SOFTWARE.
  * If a device doesn't support VAO though it claims to support should add exceptions list here.
  */
 #ifndef CC_TEXTURE_ATLAS_USE_VAO
-#define CC_TEXTURE_ATLAS_USE_VAO 1
+#if (CC_TARGET_PLATFORM == CC_PLATFORM_OHOS)
+    #define CC_TEXTURE_ATLAS_USE_VAO 0 // TBD ohos doesn't support VAO
```
Comment (dumganhar) [inline cocos/base/ccConfig.h]: VAO is a standard of OpenGLES 3.x. If ohos supports OpenGLES 3.x, it should be able to support VAO.

File: cocos/platform/CCApplicationProtocol.h [hunk 10]
```diff
@@ -58,7 +58,8 @@ class CC_DLL ApplicationProtocol
         OS_EMSCRIPTEN,  /**< Emscripten */
         OS_TIZEN,       /**< Tizen */
         OS_WINRT,       /**< Windows Runtime Applications */
-        OS_WP8          /**< Windows Phone 8 Applications */
+        OS_WP8,         /**< Windows Phone 8 Applications */
+        OS_HARMONY      /**< HARMONY */
```
Comment (dumganhar) [inline cocos/platform/CCApplicationProtocol.h]:
```suggestion
        OS_OPENHARMONY      /**< OPENHARMONY */
```

File: cocos/renderer/CCRenderer.cpp [hunk 11]
```diff
@@ -793,30 +793,33 @@ void Renderer::drawBatchedTriangles()
...
+       glBufferData(GL_ARRAY_BUFFER, sizeof(_verts[0]) * _filledVertex, nullptr, GL_STATIC_DRAW);
+       void *buf = glMapBuffer(GL_ARRAY_BUFFER, GL_WRITE_ONLY);
+       memcpy(buf, _verts, sizeof(_verts[0]) * _filledVertex);
+       glUnmapBuffer(GL_ARRAY_BUFFER);
+
+       glBindBuffer(GL_ARRAY_BUFFER, 0);
+       
+       glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, _buffersVBO[1]);
+       glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof(_indices[0]) * _filledIndex, _indices, GL_STATIC_DRAW);
+#endif
```
Comment (minggo) [inline cocos/renderer/CCRenderer.cpp:822]: It seems the content is the same, why has so many modifications?

File: cocos/renderer/CCTextureAtlas.cpp [hunk 12]
```diff
@@ -618,45 +618,47 @@ void TextureAtlas::drawNumberOfQuads(ssize_t numberOfQuads, ssize_t start)
...
+       glDrawElements(GL_TRIANGLES, (GLsizei) numberOfQuads*6, GL_UNSIGNED_SHORT, (GLvoid*) (start*6*sizeof(_indices[0])) );
+       
+       GL::bindVAO(0);
+       
 #if CC_REBIND_INDICES_BUFFER
-        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
+       glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
+#endif
 #endif
-
```
Comment (minggo) [inline cocos/renderer/CCTextureAtlas.cpp:659]: The same as above, did you modify the ident?

File: cocos/platform/ohos/CCGL-ohos.h [hunk 13]
```diff
@@ -0,0 +1,43 @@
+#ifndef __CCGL_H__
+#define __CCGL_H__
```
Comment (dumganhar) [inline cocos/platform/ohos/CCGL-ohos.h]:
```suggestion
#ifndef __CCGL_OHOS_H__
#define __CCGL_OHOS_H__
```

File: cocos/scripting/lua-bindings/script/cocos2d/Cocos2dConstants.lua [hunk 14]
```diff
@@ -188,6 +188,7 @@ cc.PLATFORM_OS_EMSCRIPTEN = 8
 cc.PLATFORM_OS_TIZEN   = 9
 cc.PLATFORM_OS_WINRT   = 10
 cc.PLATFORM_OS_WP8     = 11
+cc.PLATFORM_OS_OHOS     = 12
```
Comment (minggo) [inline cocos/scripting/lua-bindings/script/cocos2d/Cocos2dConstants.lua]: It is better to align `=`.

File: cocos/platform/ohos/CCLogOhos.h [hunk 15]
```diff
@@ -0,0 +1,121 @@
+#ifndef DISP_COMMON_H
```
Comment (dumganhar) [inline cocos/platform/ohos/CCLogOhos.h]: What's DISP_COMMON_H ?
Should it be `#ifndef __CC_LOG_OHOS_H__` ?

File: cocos/platform/ohos/CCPlatformDefine-ohos.h [hunk 16]
```diff
@@ -0,0 +1,37 @@
+#ifndef __CCPLATFORMDEFINE_H__
```
Comment (dumganhar) [inline cocos/platform/ohos/CCPlatformDefine-ohos.h]:
```suggestion
#ifndef __CCPLATFORMDEFINE_OHOS_H__
```

File: cocos/platform/ohos/CCStdC-ohos.h [hunk 17]
```diff
@@ -0,0 +1,23 @@
+#ifndef __CC_STD_C_H__
```
Comment (dumganhar) [inline cocos/platform/ohos/CCStdC-ohos.h]:
```suggestion
#ifndef __CC_STD_C_OHOS_H__
```

File: cocos/platform/ohos/CCTextBitmap.cpp [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+
+void* CCTextBitmap::getPixelAddr()
+{
+    return pixelAddr;
+}
+
+CCTextBitmap::~CCTextBitmap()
+{
+  OH_Drawing_CanvasDestroy(_canvas);
+//  OH_Drawing_BitmapDestroy(_bitmap);
```
Comment (dumganhar) [inline cocos/platform/ohos/CCTextBitmap.cpp]: Doesn't `_bitmap` need to be released?

File: cocos/scripting/js-bindings/auto/jsb_cocos2dx_audioengine_auto.cpp [hunk 19]
```diff
@@ -1,5 +1,5 @@
 #include "scripting/js-bindings/auto/jsb_cocos2dx_audioengine_auto.hpp"
-#if (CC_TARGET_PLATFORM == CC_PLATFORM_WINRT || CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID || CC_TARGET_PLATFORM == CC_PLATFORM_IOS || CC_TARGET_PLATFORM == CC_PLATFORM_MAC || CC_TARGET_PLATFORM == CC_PLATFORM_WIN32)
+#if (CC_TARGET_PLATFORM == CC_PLATFORM_OHOS || CC_TARGET_PLATFORM == CC_PLATFORM_WINRT || CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID || CC_TARGET_PLATFORM == CC_PLATFORM_IOS || CC_TARGET_PLATFORM == CC_PLATFORM_MAC || CC_TARGET_PLATFORM == CC_PLATFORM_WIN32)
```
Comment (dumganhar) [inline cocos/scripting/js-bindings/auto/jsb_cocos2dx_audioengine_auto.cpp:2]: https://github.com/cocos2d/cocos2d-x/blob/v3/tools/tojs/cocos2dx_audioengine.ini#L10C21-L10C26

This file should also be updated.

Comment (dumganhar) [reply cocos/scripting/js-bindings/auto/jsb_cocos2dx_audioengine_auto.cpp:2]: The same as other `_auto` suffix files.

File: cocos/scripting/lua-bindings/auto/lua_cocos2dx_audioengine_auto.cpp [hunk 20]
```diff
@@ -1,5 +1,5 @@
 #include "scripting/lua-bindings/auto/lua_cocos2dx_audioengine_auto.hpp"
-#if CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID || CC_TARGET_PLATFORM == CC_PLATFORM_IOS || CC_TARGET_PLATFORM == CC_PLATFORM_MAC || CC_TARGET_PLATFORM == CC_PLATFORM_WIN32 || CC_TARGET_PLATFORM == CC_PLATFORM_TIZEN
+#if CC_TARGET_PLATFORM == CC_PLATFORM_OHOS || CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID || CC_TARGET_PLATFORM == CC_PLATFORM_IOS || CC_TARGET_PLATFORM == CC_PLATFORM_MAC || CC_TARGET_PLATFORM == CC_PLATFORM_WIN32 || CC_TARGET_PLATFORM == CC_PLATFORM_TIZEN
```
Comment (dumganhar) [inline cocos/scripting/lua-bindings/auto/lua_cocos2dx_audioengine_auto.cpp:2]: Should also modify https://github.com/cocos2d/cocos2d-x/blob/v3/tools/tolua/cocos2dx_audioengine.ini#L10

File: cocos/scripting/lua-bindings/script/framework/device.lua [hunk 21]
```diff
@@ -58,6 +58,8 @@ elseif target == cc.PLATFORM_OS_WINRT then
     device.platform = "winrt"
 elseif target == cc.PLATFORM_OS_WP8 then
     device.platform = "wp8"
+elseif target == cc.PLATFORM_OS_OHOS then
+    device.platform = "harmony"
```
Comment (dumganhar) [inline cocos/scripting/lua-bindings/script/framework/device.lua]:
```suggestion
    device.platform = "openharmony"
```

File: cocos/ui/CMakeLists.txt [hunk 22]
```diff
@@ -70,6 +70,20 @@ elseif(ANDROID)
         # it's special for android, not a common file
         ui/UIWebView.cpp
         )
+ELSEIF(OHOS)
```
Comment (dumganhar) [inline cocos/ui/CMakeLists.txt]:
```suggestion
elseif(OHOS)
```

File: cocos/ui/UIVideoPlayer-ohos.h [hunk 23]
```diff
@@ -0,0 +1,5 @@
+#if (CC_TARGET_PLATFORM == CC_PLATFORM_OHOS)
```
Comment (dumganhar) [inline cocos/ui/UIVideoPlayer-ohos.h:3]: Add `#pragma once`

**Comment by dumganhar** [comment]
Why add `tests/lua-empty-test/res/farm.png` file?
There is already a `farm.jpg` in that folder. Doesn't openharmony support `jpeg decoding`?

**Comment by minggo** [comment]
About `doc/game-porting-guide`:
- it is better to named `docs/openharmony/game-porting-guide`
- please don't use chinese in folder name, such as `OpenHarmony系统架构.png`

File: extensions/assets-manager/AssetsManager.cpp [hunk 24]
```diff
@@ -197,7 +197,6 @@ std::string AssetsManager::keyOfDownloadedVersion() const
 bool AssetsManager::checkUpdate()
 {
     if (_versionFileUrl.size() == 0 || _isDownloading) return false;
-
```
Comment (minggo) [inline extensions/assets-manager/AssetsManager.cpp]: Having a space line makes codes more clear. So i don't think need to remove it.

File: tests/cpp-tests/CMakeLists.txt [hunk 25]
```diff
@@ -242,6 +249,8 @@ list(APPEND GAME_SOURCE
      Classes/ConsoleTest/ConsoleTest.cpp
      Classes/CurlTest/CurlTest.cpp
      Classes/CurrentLanguageTest/CurrentLanguageTest.cpp
+     Classes/SDKTest/SDKTest.cpp
+     Classes/PaySDKTest/PaySDKTest.cpp
```
Comment (minggo) [inline tests/cpp-tests/CMakeLists.txt]: Can these tests work on all platforms?

File: tests/cpp-tests/Classes/PaySDKTest/PaySDKTest.cpp [hunk 26]
```diff
@@ -0,0 +1,197 @@
...
+} // namespace
+
+PaySDKTests::PaySDKTests() { ADD_TEST_CASE(PaySDKControlTest); }
+
+PaySDKControlTest::PaySDKControlTest() {
+  auto label = Label::createWithTTF("Pay SDK Test", "fonts/arial.ttf", 28);
+  addChild(label, 0);
+  label->setPosition(VisibleRect::center().x, VisibleRect::top().y - 50);
+
+    //检查支付环境
```
Comment (minggo) [inline tests/cpp-tests/Classes/PaySDKTest/PaySDKTest.cpp]: Please don't use chinese in comment.

File: tests/cpp-tests/CMakeLists.txt (hunk 25 above)
Comment (minggo) [reply tests/cpp-tests/CMakeLists.txt]: And PaySDK is not part or engine, so it is not suitable to test it here.

File: tests/cpp-tests/Classes/ExtensionsTest/NetworkTest/WebSocketDelayTest.cpp [hunk 27]
```diff
@@ -100,7 +100,7 @@ void WebSocketDelayTest::startTestCallback(Ref* sender)
     std::vector<std::string> protocols;
     protocols.push_back("myprotocol_1");
     protocols.push_back("myprotocol_2");
-    if (!_wsiSendText->init(*this, "wss://echo.websocket.org", &protocols, "cacert.pem"))
+    if (!_wsiSendText->init(*this, "ws://echo.websocket.events", &protocols, "cacert.pem"))
```
Comment (minggo) [inline tests/cpp-tests/Classes/ExtensionsTest/NetworkTest/WebSocketDelayTest.cpp:103]: Why modify it?

File: tests/cpp-tests/Classes/ExtensionsTest/NetworkTest/WebSocketTest.cpp [hunk 28]
```diff
@@ -137,7 +137,8 @@ void WebSocketTest::startTestCallback(Ref* sender)
     std::vector<std::string> protocols;
     protocols.push_back("myprotocol_1");
     protocols.push_back("myprotocol_2");
-    if (!_wsiSendText->init(*this, "wss://echo.websocket.org", &protocols, "cacert.pem"))
+    // if (!_wsiSendText->init(*this, "wss://echo.websocket.events", &protocols, "cacert.pem"))
+    if (!_wsiSendText->init(*this, "ws://echo.websocket.events"))
```
Comment (minggo) [inline tests/cpp-tests/Classes/ExtensionsTest/NetworkTest/WebSocketTest.cpp:141]: The same as above.

**Comment by minggo** [comment]
Why the files in `tests/cpp-tests/proj.ohos/dependencies`, such as `hvigor-3.0.4-s.tgz`,  needed?

File: extensions/Particle3D/PU/CCPUMaterialManager.cpp (hunk 1 above)
Comment (1-newbie) [reply extensions/Particle3D/PU/CCPUMaterialManager.cpp]: yes，string，we modified it，will submit

File: cmake/Modules/CocosBuildSet.cmake (hunk 3 above)
Comment (1-newbie) [reply cmake/Modules/CocosBuildSet.cmake]: We need to use luatests and jstests to test the functionality of the engine, but the default value is off, making each modification more cumbersome. Therefore, we choose to keep this switch on. The next update of PR will close it

File: cocos/platform/ohos/CCApplication-ohos.cpp (hunk 4 above)
Comment (1-newbie) [reply cocos/platform/ohos/CCApplication-ohos.cpp:1]: CCApplication-android.h has the similar macro, we are just trying to maintain consistency in style

File: cocos/platform/ohos/CCApplication-ohos.h (hunk 5 above)
Comment (1-newbie) [reply cocos/platform/ohos/CCApplication-ohos.h:5]: CCApplication-android.h has the similar macro, we are just trying to maintain consistency in style

File: cocos/audio/include/AudioEngine.h (hunk 8 above)
Comment (1-newbie) [reply cocos/audio/include/AudioEngine.h]: will delete

File: cocos/base/ccConfig.h (hunk 9 above)
Comment (1-newbie) [reply cocos/base/ccConfig.h]: yes, ohos supports GLES3, but not support GLES2 yet. if we set 1 here, it will allow to execute the follow code. Therefore, we turned off support for VAO and blocked the following code through macros
image

File: cocos/renderer/CCRenderer.cpp (hunk 11 above)
Comment (1-newbie) [reply cocos/renderer/CCRenderer.cpp:822]: we just add macro to avoid compile error for not support GLES2, no idea to why has so many modifications

Comment (1-newbie) [reply cocos/renderer/CCRenderer.cpp:822]: we just add macro to avoid compile error for not support GLES2. It seems the indent caused so many modifications, we will restore it

File: cocos/renderer/CCTextureAtlas.cpp (hunk 12 above)
Comment (1-newbie) [reply cocos/renderer/CCTextureAtlas.cpp:659]: ok, will recover

File: cocos/scripting/lua-bindings/script/cocos2d/Cocos2dConstants.lua (hunk 14 above)
Comment (1-newbie) [reply cocos/scripting/lua-bindings/script/cocos2d/Cocos2dConstants.lua]: ok

File: cocos/platform/ohos/CCLogOhos.h (hunk 15 above)
Comment (1-newbie) [reply cocos/platform/ohos/CCLogOhos.h]: ok, modified, will submit

File: cocos/scripting/js-bindings/auto/jsb_cocos2dx_audioengine_auto.cpp (hunk 19 above)
Comment (1-newbie) [reply cocos/scripting/js-bindings/auto/jsb_cocos2dx_audioengine_auto.cpp:2]: image

Are you saying we need modify all the files under tojs and tolua like this?

Comment (1-newbie) [reply cocos/scripting/js-bindings/auto/jsb_cocos2dx_audioengine_auto.cpp:2]: We haven't used this tool, so we haven't focused on this part before.

File: cocos/ui/UIVideoPlayer-ohos.h (hunk 23 above)
Comment (1-newbie) [reply cocos/ui/UIVideoPlayer-ohos.h:3]: added

File: cocos/base/ccConfig.h (hunk 9 above)
Comment (dumganhar) [reply cocos/base/ccConfig.h]: So you mean that there is not a `glMapBufferOES` extension in GLES2 implementation on OH platform?

Comment (dumganhar) [reply cocos/base/ccConfig.h]: 企业微信截图_02435cc0-c72c-4384-bca4-e2727e3d18df

File: cocos/scripting/js-bindings/auto/jsb_cocos2dx_audioengine_auto.cpp (hunk 19 above)
Comment (dumganhar) [reply cocos/scripting/js-bindings/auto/jsb_cocos2dx_audioengine_auto.cpp:2]: Yes. `*_auto.cpp` files are generated by `tolua/tojs` tools automatically.

File: extensions/assets-manager/AssetsManager.cpp (hunk 24 above)
Comment (1-newbie) [reply extensions/assets-manager/AssetsManager.cpp]: recovered

File: tests/cpp-tests/CMakeLists.txt (hunk 25 above)
Comment (1-newbie) [reply tests/cpp-tests/CMakeLists.txt]: ok, deleted

File: tests/cpp-tests/Classes/PaySDKTest/PaySDKTest.cpp (hunk 26 above)
Comment (1-newbie) [reply tests/cpp-tests/Classes/PaySDKTest/PaySDKTest.cpp]: deleted

File: tests/cpp-tests/Classes/ExtensionsTest/NetworkTest/WebSocketDelayTest.cpp (hunk 27 above)
Comment (1-newbie) [reply tests/cpp-tests/Classes/ExtensionsTest/NetworkTest/WebSocketDelayTest.cpp:103]: I remember the orignal one doesn't work

File: cocos/audio/AudioEngine.cpp (hunk 2 above)
Comment (1-newbie) [reply cocos/audio/AudioEngine.cpp]: recovered

File: cocos/audio/AudioEngine.cpp (hunk 6 above)
Comment (1-newbie) [reply cocos/audio/AudioEngine.cpp]: recovered

File: cocos/audio/AudioEngine.cpp (hunk 7 above)
Comment (1-newbie) [reply cocos/audio/AudioEngine.cpp]: recovered

File: cocos/base/ccConfig.h (hunk 9 above)
Comment (1-newbie) [reply cocos/base/ccConfig.h]: yes, no gles2 on ohos now. this header file is pre-added.
in the cmakelist, no link to GLES2 now

File: cocos/platform/ohos/CCTextBitmap.cpp (hunk 18 above)
Comment (1-newbie) [reply cocos/platform/ohos/CCTextBitmap.cpp]: the destructor function of CCTextBitmap seems never called. we fixed this bug

File: cocos/scripting/js-bindings/auto/jsb_cocos2dx_audioengine_auto.cpp (hunk 19 above)
Comment (1-newbie) [reply cocos/scripting/js-bindings/auto/jsb_cocos2dx_audioengine_auto.cpp:2]: ok

File: tests/cpp-tests/Classes/ExtensionsTest/NetworkTest/WebSocketTest.cpp (hunk 28 above)
Comment (1-newbie) [reply tests/cpp-tests/Classes/ExtensionsTest/NetworkTest/WebSocketTest.cpp:141]: the orignal one doesn't work

File: cocos/base/CCController-ohos.cpp [hunk 29]
```diff
@@ -0,0 +1,164 @@
...
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ ****************************************************************************/
+
+#include "base/CCController.h"
+
+#if (CC_TARGET_PLATFORM == CC_PLATFORM_OHOS)
```
Comment (qiuguohua) [inline cocos/base/CCController-ohos.cpp:29]: This is a file unique to ohos and should not require the addition of macro definitions?

File: cocos/platform/CCCommon.h [hunk 30]
```diff
@@ -37,6 +37,9 @@ NS_CC_BEGIN
  * @{
  */
 
+/// The max length of CCLog message.
+static const int kMaxLogLen = 16*1024;
```
Comment (qiuguohua) [inline cocos/platform/CCCommon.h]: What is this definition used for?

File: cocos/platform/ohos/CCApplication-ohos.cpp [hunk 31]
```diff
@@ -0,0 +1,114 @@
...
+Application::~Application()
+{
+    CCAssert(this == sm_pSharedApplication, "");
+    sm_pSharedApplication = nullptr;
+}
+
+int Application::run()
+{
+    // Initialize instance and cocos2d.
+    if (! applicationDidFinishLaunching())
```
Comment (qiuguohua) [inline cocos/platform/ohos/CCApplication-ohos.cpp]: extra space

File: cocos/platform/ohos/CCDevice-ohos.cpp [hunk 32]
```diff
@@ -0,0 +1,112 @@
...
+    {
+        CCTextBitmap *cCtextBitmap = new CCTextBitmap();
+        CCTextBitmap::createCCTextBitmap(cCtextBitmap, text, textDefinition._fontName.data(), textDefinition._fontAlpha, textDefinition._fontFillColor.r, 
+            textDefinition._fontFillColor.g, textDefinition._fontFillColor.b ,eAlignMask, nWidth, nHeight, textDefinition._fontSize);
+        void* pixels = cCtextBitmap->getPixelAddr();
+        cocos2d::BitmapDC& bitmapDC = sharedBitmapDC();
+        bitmapDC.m_nWidth = cCtextBitmap->GetWidth();
+        bitmapDC.m_nHeight = cCtextBitmap->GetHeight();
+        long size = bitmapDC.m_nWidth * bitmapDC.m_nHeight * 4;
+        bitmapDC.m_pData = new unsigned char[size];
```
Comment (qiuguohua) [inline cocos/platform/ohos/CCDevice-ohos.cpp]: sharedBitmapDC is a static variable and will be freed when the program exits, if it is reused, there will be memory leaks

File: cocos/base/CCController-ohos.cpp (hunk 29 above)
Comment (1-newbie) [reply cocos/base/CCController-ohos.cpp:29]: just keep the code style, cause android has the same macro

Comment (qiuguohua) [reply cocos/base/CCController-ohos.cpp:29]: Okay

**Comment by 1-newbie** [comment]
> Why the files in `tests/cpp-tests/proj.ohos/dependencies`, such as `hvigor-3.0.4-s.tgz`, needed?

cause the deveco still in beta, the developer can't download hvigor from the remote repo, so we need use the local file. After the hvigor can download, we will remove these files.

File: cocos/platform/ohos/CCDevice-ohos.cpp (hunk 32 above)
Comment (1-newbie) [reply cocos/platform/ohos/CCDevice-ohos.cpp]: the m_pData will be returned as part of Data in the getTextureDataForText function and might be freed at somewhere.
in the similar function of android, it also malloc new memory every time.
final, we use Profile(IDE tools) to detected the memory usage, no noticeable memory leaks issues found

**Review by minggo — APPROVED** [review]

File: cocos/platform/ohos/CCApplication-ohos.cpp (hunk 31 above)
Comment (1-newbie) [reply cocos/platform/ohos/CCApplication-ohos.cpp]: ok, removed

File: cocos/platform/CCCommon.h (hunk 30 above)
Comment (1-newbie) [reply cocos/platform/CCCommon.h]: deleted

**Review by qiuguohua — APPROVED** [review]

File: cocos/platform/ohos/CCDevice-ohos.cpp [hunk 33]
```diff
@@ -0,0 +1,112 @@
...
+    {
+        CCTextBitmap *cCtextBitmap = new CCTextBitmap();
+        CCTextBitmap::createCCTextBitmap(cCtextBitmap, text, textDefinition._fontName.data(), textDefinition._fontAlpha, textDefinition._fontFillColor.r, 
+            textDefinition._fontFillColor.g, textDefinition._fontFillColor.b ,eAlignMask, nWidth, nHeight, textDefinition._fontSize);
+        void* pixels = cCtextBitmap->getPixelAddr();
+        cocos2d::BitmapDC& bitmapDC = sharedBitmapDC();
+        bitmapDC.m_nWidth = cCtextBitmap->GetWidth();
+        bitmapDC.m_nHeight = cCtextBitmap->GetHeight();
+        long size = bitmapDC.m_nWidth * bitmapDC.m_nHeight * 4;
+        bitmapDC.m_pData = (unsigned char*)malloc(sizeof(unsigned char) * size);
```
Comment (dumganhar) [inline cocos/platform/ohos/CCDevice-ohos.cpp:58]: m_pData is released by `delete [] m_pData;` in `~BitmapDC`. Why change to `malloc` ?

**Review by dumganhar — APPROVED** [review]
